sel/src/Sel/SecretKey/EncryptedStream.hs

```diff
@@ -0,0 +1,542 @@
+--
+-- @since 0.0.1.0
+data StreamTag
+  = -- | Most common tag, add no information about the nature of the message
+    Message
+  | -- | Indicates that the message marks the end of a set of messages, but not the end of the stream
+    Push
+  | -- | "forget" the secret key used to encrypt this message and the previous ones, and derive a new secret key.
+    Rekey
+  | -- | Marks the end of the stream, and erases the secret key used to encrypt the previous sequence.
+    Final
+  deriving
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Enum
+      -- ^ @since 0.0.1.0
+    , Bounded
+      -- ^ @since 0.0.1.0
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+    )
+    via (ShowInstance StreamTag)
+
+-- == Multipart ==
+
+-- | The cryptographic state necessary for streaming operations.
+--
+-- @since 0.0.1.0
+newtype Multipart s = Multipart (Ptr CryptoSecretStreamXChaCha20Poly1305State)
+
+type role Multipart nominal
+
+-- | This record holds the result of a decryption on a stream message
+--
+-- @since 0.0.1.0
+data StreamResult = StreamResult
+  { streamMessage :: StrictByteString
+  , mStreamTag :: Maybe StreamTag
+  , mAdditionalData :: Maybe StrictByteString
+  }
+  deriving
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+    )
+    via (ShowInstance StreamResult)
+
+-- | Possible errors that can happen during the encryption and decryption process
+--
+-- @since 0.0.1.0
+data EncryptedStreamError
+  = InvalidHeader
+  | InvalidCipherText
+  | InputStreamInitError
+  | InputStreamPushError
+  deriving
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+    )
+    via (ShowInstance EncryptedStreamError)
+
+-- == Encryption ==
+
+-- | Initialise a stream to which you will push encrypted message.
+-- This function returns the 'Header' and the 'SecretKey' that are needed for the peer to
+-- decrypt the stream.
+--
+-- @since 0.0.1.0
+initPushStream
+  :: Multipart s
+  -- ^ Cryptographic state
+  -> IO (Either EncryptedStreamError (Header, SecretKey))
+initPushStream (Multipart statePtr) = do
+  headerForeignPtr <- liftIO $ Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305HeaderBytes)
+  (SecretKey secretKeyForeignPtr) <- newSecretKey
+  Foreign.withForeignPtr headerForeignPtr $ \headerPtr ->
+    Foreign.withForeignPtr secretKeyForeignPtr $ \keyPtr -> do
+      result <-
+        cryptoSecretStreamXChaCha20Poly1305InitPush
+          statePtr
+          headerPtr
+          keyPtr
+      case result of
+        0 -> pure $ Right (Header headerForeignPtr, SecretKey secretKeyForeignPtr)
+        _ -> pure $ Left InputStreamInitError
```

Comment:
This leaks memory, unless `mallocForeignPtrBytes` automagically installs a finalizer.

Reply:
Actually yes! https://hackage.haskell.org/package/base-4.18.0.0/docs/Foreign-ForeignPtr.html#v:mallocForeignPtr

Reply:
Huh, good to know.